pkg/retrieval/retrieval_test.go

```diff
@@ -0,0 +1,130 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package retrieval_test
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io/ioutil"
+	"testing"
+	"time"
+
+	"github.com/ethersphere/bee/pkg/logging"
+	"github.com/ethersphere/bee/pkg/p2p/mock"
+	"github.com/ethersphere/bee/pkg/p2p/protobuf"
+	"github.com/ethersphere/bee/pkg/retrieval"
+	storemock "github.com/ethersphere/bee/pkg/storage/mock"
+)
+
+type mockPeerSuggester struct {
+	returnValue string
+}
+
+func (v mockPeerSuggester) SuggestPeer(_ []byte) (string, error) {
+	return v.returnValue, nil
+}
+
+// TestDelivery tests that a naive request -> delivery flow works
+func TestDelivery(t *testing.T) {
+	logger := logging.New(ioutil.Discard)
+
+	mockStorer := storemock.NewStorer()
+	reqAddr := []byte("reqaddr")
+	reqData := []byte("data data data")
+
+	// put testdata in the mock store of the server
+	_ = mockStorer.Put(context.TODO(), reqAddr, reqData)
+
+	// create the server that will handle the request and will serve the response
+	server := retrieval.New(retrieval.Options{
+		Storer: mockStorer,
+		Logger: logger,
+	})
+	recorder := mock.NewRecorder(
+		mock.WithProtocols(server.Protocol()),
+	)
+
+	// client mock storer does not store any data at this point
+	// but should be checked at at the end of the test for the
+	// presence of the reqAddr key and value to ensure delivery
+	// was successful
+	clientMockStorer := storemock.NewStorer()
+
+	ps := mockPeerSuggester{returnValue: "/p2p/QmZt98UimwpW9ptJumKTq7B7t3FzNfyoWVNGcd8PFCd7XS"}
+	client := retrieval.New(retrieval.Options{
+		Streamer:      recorder,
+		PeerSuggester: ps,
+		Storer:        clientMockStorer,
+		Logger:        logger,
+	})
+
+	errc := make(chan error)
+	go func() {
```

Comment:
Could the timeout be set on the context to avoid creating a goroutine, errc and select to handle timeout?

Reply:
i committed this change but it deadlocks. are libp2p streams handled synchronously? if so - unless we make the handling async there might be no way around creating a new goroutine

Reply:
Yes, you are right. Recorder is using buffers that are synchronous. I am sorry, your previous code was correct. Would you mind putting it back as there is a warning that Fatalf is called in a goroutine?